Adding in the single use link model, generation controller, and view controller

cam156 opened the pull request:
see #146 

This code is copied from Sufia.  The controllers, models, and views for generating and showing sungle use links are in this PR.  The javascript used in sufia to generate the link in the dashboard and copy it to the user clipboard is not.

grosscol commented on app/controllers/curation_concerns/single_use_links_viewer_controller.rb:
```diff
@@ -0,0 +1,21 @@
+
+        can :read, ActiveFedora::Base do |obj|
+          single_use_link.valid? && single_use_link.itemId == obj.id && single_use_link.destroy!
+        end if single_use_link
```
This looks like a rubocop-ism.

cam156 replied:
@grosscol I do not think it is a mistake.  It means only run the block if single_use_link is defined.

mjgiarlo replied:
My $0.02is not to sweat the rubocopisms 'til #152 is merged (not ready yet). I'll rebase on master and take care of all offenses in one fell swoop, so no worries here.

grosscol commented on app/views/layouts/error.html.erb:
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
+<html lang="<%= t("sufia.document_language") %>">
```
These localization references need to be curation_concerns.  Also, I couldn't find the localization string document_language.

cam156 replied:
@grosscol Good catch.  I'll fix these up.

grosscol commented on spec/controllers/curation_concerns/single_use_links_viewer_controller_spec.rb:
```diff
@@ -0,0 +1,73 @@
+  routes { CurationConcerns::Engine.routes }
+  let(:file) do
+    GenericFile.create do |file|
+      file.add_file(File.open(fixture_path + '/world.png'), path: 'content', original_name: 'world.png', mime_type: 'image/png')
```
Should this be Hydra::Works::AddFileToGenericFile?  I couldn't find where add_file gets define or comes from.

cam156 replied:
@grosscol it is defined here: https://github.com/projecthydra/hydra-head/blob/master/hydra-core/app/models/concerns/hydra/model_methods.rb#L16-L22
I'm not sure what methods exists in hydra works, so you may be correct.

cam156 commented:
@grosscol Using the Hydra::Works showed me I was using the incorrect download module, so very good catch.  I believe it is all fixed up.

mjgiarlo commented:
I'd like one or more of the Princetonians to review this before I :shipit: 

Thanks, @cam156!

mjgiarlo commented:
Even future Princetonians. ;)  @terrellt @escowles @tampakis @jpstroop @kevinreiss

tampakis commented:
@mjgiarlo I got it.

tampakis commented on app/views/layouts/error.html.erb:
```diff
@@ -0,0 +1,34 @@
+    <!-- application js -->
+    <%= javascript_include_tag 'application' %>
+
+    <%= render partial: '/ga', formats: [:html] %>
```
'shared/ga'

tampakis commented on app/views/layouts/error.html.erb:
```diff
@@ -0,0 +1,34 @@
+<div  id="wrapper">
+  <div class="container-fluid">
+    <div id="page-positioner">
+      <%= render partial: '/masthead', formats: [:html] %>
```
Remove this line. Curation Concerns doesn't have a masthead partial.

tampakis commented on app/views/layouts/error.html.erb:
```diff
@@ -0,0 +1,34 @@
+            </div>
+        </div><!-- /#content -->
+      </div><!-- /#content-wrapper -->
+      <%= render partial: '/footer' %>
```
'shared/footer'

cam156 commented:
@tampakis Thanks for the review.  I will fix up what you found.

tampakis commented on app/views/curation_concerns/single_use_links_viewer/show.html.erb:
```diff
@@ -0,0 +1,21 @@
+  <div class="well active" id="descriptions_display">
+    <table class="table table-striped">
+      <tbody>
+      <% present_terms(@presenter) do |r, term| %>
```
Curation Concerns doesn't use the presenter to render the Generic File metadata. Lines 7 through 21 can be replaced with:
```
<%= render "curation_concerns/base/attributes", curation_concern: @asset %>
```

tampakis replied:
This bit here at line 12 is what doesn't work because of the missing `present_terms` [helper method](https://github.com/projecthydra/sufia/blob/a064489d3cf4d228abf978d7126644e2c653aa5e/app/helpers/generic_file_helper.rb#L7), which calls the missing `PresenterRenderer` [class](https://github.com/projecthydra/sufia/blob/a064489d3cf4d228abf978d7126644e2c653aa5e/app/presenters/sufia/presenter_renderer.rb).

tampakis commented on app/controllers/concerns/curation_concerns/single_use_links_viewer_controller_behavior.rb:
```diff
@@ -0,0 +1,64 @@
+      rescue_from CanCan::AccessDenied, with: :render_single_use_error
+      rescue_from ActiveRecord::RecordNotFound, with: :render_single_use_error
+      class_attribute :presenter_class
+      self.presenter_class = CurationConcerns::GenericFilePresenter
```
I think that the presenter-based stuff (L12, 13, 25, 34-36) can be removed since it's not needed for the the single-use show view.

tampakis commented on app/views/curation_concerns/single_use_links/new_download.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div class="single-use-link">
+<h1>Single Use Link</h1>
+<p>Anyone can use the following link once to download the file</p>
+<%= link_to @asset, @link, class: 'download-link' %> <%= link_to raw('<i class="glyphicon glyphicon-link large-icon"></i>'), '#', class: 'copypaste itemicon itemcode', title: 'Copy File URL', id: "copy_link_#{@asset.id}" %>
```
The `link_to_raw` bit refers to javascript code that isn't there so that bit should be removed. The download links only work when they are opened in a new tab so I think ` target: '_blank'` should be provided as an option after `class: 'download-link'`

tampakis commented on app/views/curation_concerns/single_use_links_viewer/show.html.erb:
```diff
@@ -0,0 +1,21 @@
+  <h1 class="lower"><%= @asset %></h1>
+    <h2 class="non lower">Actions</h2>
+    <p>
+      <%= link_to "Download (can only be used once)", @download_link %>
```
` target: '_blank'` here as well to prevent the error page from showing.

tampakis commented:
@cam156 I added a few more comments. I'll merge when they've been addressed. This looks great. Thanks!

cam156 commented:
@tampakis Are we sure we do not want to head down the presenter route?  Those were introduced relatively recently to sufia and people seem to like them.  I was not around for the conversations you had so I'm good using the base view, just double checking where we are heading. @mjgiarlo @jpstroop you may have some insight into presenter or no presenter.

jcoyne commented:
@tampakis Presenters are very useful, especially in the case that someone wants to customize the metadata. Furthermore they simplify the interface between the view and the controller.  I've got a very big :+1: for this.

mjgiarlo commented:
:+1: to using presenters.

tampakis commented:
The presenters do look cool. It's just that they aren't fully implemented in CC. It doesn't have the `present_terms` [helper method](https://github.com/projecthydra/sufia/blob/a064489d3cf4d228abf978d7126644e2c653aa5e/app/helpers/generic_file_helper.rb#L7) or a `PresenterRenderer` [class](https://github.com/projecthydra/sufia/blob/a064489d3cf4d228abf978d7126644e2c653aa5e/app/presenters/sufia/presenter_renderer.rb). I thought it would be easier to use the already adequate template partial than to refactor generic_file, work, and collection to use presenters at this point.

jcoyne commented:
The `present_terms` and `PresenterRenderer` are for the automatic conversion of the presenter into an HTML form. But that feature is in no way necessary to effectively use presenters.  Presenters live at the View level, so there is no need to refactor anything at the model level (unless someone put view logic there that should be moved into a more appropriate place)

tampakis commented:
I'm just saying that if the presenter is used here for the single-use view, it should be applied to the other views as well.

jcoyne commented:
@tampakis the journey of a thousand miles begins with a single step.  You can't require someone to refactor the whole app at one time in order to get a little pull request done. If that were a rule, there would be no hydra.

jcoyne commented:
@cam156 can you please add a test for `app/views/curation_concerns/single_use_links_viewer/show.html.erb`  It doesn't look like it'll work (calling a non-existant method)

mjgiarlo commented:
Perhaps we can resolve the difference of opinions about presenters by 1) starting with *this* presenter in the PR, and 2) ticketing the creation of other presenters so that CC is consistent? cc: @jcoyne @tampakis

tampakis commented:
:+1:

mjgiarlo commented:
:clap: 

@tampakis That @jpstroop guy hinted that you might be inclined to put in those tickets. True?

tampakis commented on app/views/curation_concerns/single_use_links/new_show.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div class="single-use-link">
+<h1>Single Use Link</h1>
+<p>Anyone can use the following link once to view the file metadata</p>
+<%= link_to @asset, @link %>
```
@cam156 for some reason this link also gives an error without including `target: '_blank'` when using the presenter class.

cam156 commented:
@tampakis @jcoyne @mjgiarlo and others...

Added tests for the views, which showed the error @jcoyne was expecting.  Added in a minimum set of fields for the presenter and got all the test running.

jcoyne commented:
Can we please merge #200 before this one? These two PRs are going to clash real hard and I think it'll be easier to rebase #151 on top of #200 instead of vice versa.  I'll volunteer to do the rebase if @cam156 doesn't want to.

cam156 commented:
@jcoyne This one has already been rebased 3 times.  Thats the issue with having a PR open for two weeks.  Unfortunately I will not have any time to rebase it again today, so if the other item gets merged @jcoyne go for it.

mjgiarlo commented:
If @jcoyne is offering to rebase this, I'll take him up on that. ;)

jcoyne commented on app/controllers/concerns/curation_concerns/single_use_links_viewer_controller_behavior.rb:
```diff
@@ -0,0 +1,63 @@
+    end
+
+    def show
+      raise not_found_exception unless single_use_link.path == polymorphic_path([main_app, :curation_concerns, @asset])
```
How does `@asset` get loaded on this line?

jcoyne replied:
Nevermind, must be download_behavior

jcoyne commented:
@mjgiarlo rebased.

cam156 commented:
@jcoyne Thanks! @mjgiarlo @tampakis Good to merge?

tampakis commented:
Looks good to me.
